From a8667828d639793e80a18cee43b0d6a23573cd53 Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Wed, 11 Oct 2023 16:57:07 +0200
Subject: [PATCH 0298/1695] core: add platform-specific abort handler

Platforms may have specific code to handle an abort  when fault type
is FAULT_TYPE_IGNORE. Add plat_abort_handler() that should be
overloaded in platform code.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 72df9e84283b8e2c37c6eb72361202d89b04ff77 core: add platform-specific abort handler]
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Change-Id: I2a701d378b9d7f9533e8ce22cd5d319dd4bef7de
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/333124
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 core/arch/arm/kernel/abort.c | 11 ++++++++++-
 core/include/kernel/abort.h  |  4 ++++
 2 files changed, 14 insertions(+), 1 deletion(-)

diff --git a/core/arch/arm/kernel/abort.c b/core/arch/arm/kernel/abort.c
index 72fbfc84f..4adbc99fb 100644
--- a/core/arch/arm/kernel/abort.c
+++ b/core/arch/arm/kernel/abort.c
@@ -548,15 +548,24 @@ static enum fault_type get_fault_type(struct abort_info *ai)
 	}
 }
 
+void __weak plat_abort_handler(struct thread_abort_regs *regs __unused)
+{
+	/* Do nothing */
+}
+
 void abort_handler(uint32_t abort_type, struct thread_abort_regs *regs)
 {
 	struct abort_info ai;
+	enum fault_type type;
 	bool handled;
 
 	set_abort_info(abort_type, regs, &ai);
 
-	switch (get_fault_type(&ai)) {
+	type = get_fault_type(&ai);
+	switch (type) {
 	case FAULT_TYPE_IGNORE:
+		/* Allow platform-specific handling */
+		plat_abort_handler(regs);
 		break;
 	case FAULT_TYPE_USER_MODE_PANIC:
 		DMSG("[abort] abort in User mode (TA will panic)");
diff --git a/core/include/kernel/abort.h b/core/include/kernel/abort.h
index 4a0f5270d..286e29c18 100644
--- a/core/include/kernel/abort.h
+++ b/core/include/kernel/abort.h
@@ -27,11 +27,15 @@ struct abort_info {
 
 /* Print abort info to the console */
 void abort_print(struct abort_info *ai);
+
 /* Print abort info + stack dump to the console */
 void abort_print_error(struct abort_info *ai);
 
 void abort_handler(uint32_t abort_type, struct thread_abort_regs *regs);
 
+/* Platform overload, should be implemented in platform code */
+void plat_abort_handler(struct thread_abort_regs *regs);
+
 bool abort_is_user_exception(struct abort_info *ai);
 
 bool abort_is_write_fault(struct abort_info *ai);
-- 
2.47.2

